### DISCUSSION TITLE
Fastify plugin child logger?

### DISCUSSION THREAD

**Discussion by AviBueno** [fastify/fastify#2663] [Q&A]
Have a few plugins and I'd like to create child loggers for each one of them in order to redact plugin-related pieces of information.
Having a hard time finding info on the subject.
Is there a way to achieve that?

**Comment by jsumners** [comment]
No. We don't currently have a mechanism for that which would cover everything. You could create a plugin that creates a local child logger like `const myLogger = fastify.log.child({ namespace: 'foo' })`, but that wouldn't help with request scoped logs. I think a PR would be seriously considered for this feature.

**Comment by kibertoad** [comment]
@AviBueno If you need request-scoped logs, you should consider using https://www.npmjs.com/package/fastify-request-context

**Reply by AviBueno** [reply]
@kibertoad thanks for introducing the plugin. If I understand correctly, this seems to be good way to enrich a specific request without the need to augment Fastify's global Request object per each individual property.

I can see other use cases for it, but I'm not sure this one qualifies because I'd like to override the content of `req.log` which is an already existing property (see my reply to jsumners).

**Comment by AviBueno** [comment]
@jsumners following your reply I figured that the best I can do for now will be to introduce a child logger upon `onRequest`, e.g.:

    fastify.addHook('onRequest', async (req, rep) => {
      req.log = rep.log = req.log.child({
        redact: [...]
      });
    });

* One minor issue is that when using auto-logging, the child log will kick in only after `incoming request` has been logged.
* However, the overall plot thickens because, evidently, [child logger redaction is not currently supported](https://github.com/pinojs/pino/issues/831) :pensive:

**Reply by jsumners** [reply]
Redaction is applied from the parent logger's configuration.

**Reply by AviBueno** [reply]
I understand that. In the current discussion I was looking at a way to modify that behavior at the FastifyPlugin level, reason being that each plugin has a different set of fields that should be redacted and I'd like to avoid polluting the global Fastify object with irrelevant fields between plugins.

**Reply by mcollina** [reply]
This needs some new machinery to make it all work, but I think it is possible:

1. update pino to make custom redaction in child loggers possible 
2. add some optional callback/hook to have the child logger arguments computed dynamically
3. make said hook encapsulated and part of `Context`

**Reply by AviBueno** [reply]
Pino child loggers is key, and that's not supported currently.
But regardless, a child logger is not necessarily such a common scenario so I'm not sure if a dedicated hook is required.

Placing the hook and/or the logger as part of `Context` is beyond the scope of my understanding, i.e.:
* Is `Context` = fastify-request-context or something else?
* Why should the hook be placed as part of the context and how will it get triggered that way?
* Why should the child logger be placed as part of the context (as suggested above) instead of just replacing `request.log = request.log.child(...)`?

In any case, the discussion is just theoretical right now as the whole issue is blocked on Pino child loggers, so there's no need to reply in case this discussion is a waste of time.

Thank you all for your comments.

**Reply by kibertoad** [reply]
@AviBueno Child loggers are supported (https://github.com/pinojs/pino/blob/master/docs/child-loggers.md), child logger redaction is not :)
